### ISSUE TITLE
Generate JSON Schema output

### ISSUE THREAD

**Issue by voxpelli** [asyncapi/modelina#532]
#### Reason/Context

As JSON Schema is a common denominator of all of the input types, being able to take any type of input and output normalized JSON Schema would make this tool useful in other flows based on JSON Schemas.

It can then easily be plugged in to generate JSON Schema from OpenAPI/Swagger or AsyncAPI inputs + possibly even update outdated JSON Schema syntax into the latest and greatest spec version.

An additional benefit could be internally: That it would become possible to validate whether any data is lost in the conversions. Sending a JSON Schema as an input, generating a JSON Schema as an output and checking for substantial differences between the two. A good integration test?

#### Description

A new JSON Schema generator would have to be added.

This wouldn't be a breaking change.

**Comment by jonaslagoni** [comment]
Interesting, thanks for raising the issue! I agree it would be great to have. 

> An additional benefit could be internally: That it would become possible to validate whether any data is lost in the conversions. Sending a JSON Schema as an input, generating a JSON Schema as an output and checking for substantial differences between the two. A good integration test?

Yes, good point.

I don't think there will be a way to guarantee the output JSON Schema will validate the very same data as the input JSON Schema.

This is because we interpret it from validation rules (which are or can be very strict) to a looser format that only represents the underlying structure. For example, if you have the validation rule such as:

```json 
{type: "string", format: "email"}
```
For model generation (at least how it is now) we only care it has the `string` type. As email formats are the restriction on the value, not the type. Maybe in the future, we can support such restrictions 🤔 Related to https://github.com/asyncapi/modelina/issues/130

**Comment by voxpelli** [comment]
A complementary feature then could be to take eg TypeScript definitions as input as well (probably a bit complicated) and/or to add JTD as a format for both input and output: https://ajv.js.org/guide/schema-language.html#json-type-definition

**Comment by jonaslagoni** [comment]
> A complementary feature then could be to take eg TypeScript definitions as input as well (probably a bit complicated) and/or to add JTD as a format for both input and output: https://ajv.js.org/guide/schema-language.html#json-type-definition

Absolutely, please create another feature request for it, thanks!

Related TS input issue: https://github.com/asyncapi/modelina/issues/187

**Comment by ron-debajyoti** [comment]
@jonaslagoni let me work on this since this looks related to #187

**Comment by ron-debajyoti** [comment]
Hey @jonaslagoni since we're going to create a new JSON generator, it has to extend `AbstractGenerator` and the underlying abstract functions right ? So for starters, I am not thinking of adding preset support or anything right now, just generating a JSON Schema file from the `CommonInputModel`. 

Also how are we differentiating between `generateCompleteModels()` and `generate()` for this case ? 

Please point out if I have misunderstood anything.

**Comment by jonaslagoni** [comment]
> JSON generator, it has to extend AbstractGenerator and the underlying abstract functions right ?

Yes. An important difference between existing generators and a JSON Schema one, is that you only need one, regardless of whether it's an [enum, object](https://github.com/asyncapi/modelina/blob/10aca3b10d985596241761f8976697a72dc360e1/src/generators/typescript/TypeScriptGenerator.ts#L85).

You also don't have any formatting of property names or reserved keywords as everything is wrapped in `"`, so on that point, it is easier to implement 🙂 

You could take a look at one of the existing PR's that are adding a generator, both examples can be followed in terms of what is needed to add another generator 🙂 

> Also how are we differentiating between generateCompleteModels() and generate() for this case ?

Good question, I would probably suggest that we use `generate` to generate individual JSON Schema documents and `generateCompleteModels` could be seen as a bundled schema? 🤔 

However, I would only focus on `generate` and forget about `generateCompleteModels` 🙂 

How does that sound? 🤔

**Comment by MattKintonCinch** [comment]
Is there any update on this? We're trying to use Modelina to resolve all the refs within our schemas, then output them all back to us so we can write them in a flat folder structure. This would be really useful for that I think.

@jonaslagoni at the moment we're just using the `ConstrainedMetaModel` output from the TS Generator which gives us the correct dereferenced models. Is there a better way of doing this?

**Comment by jonaslagoni** [comment]
@MattKintonCinch so, you have a x nested JSON Schema file structures, that you want to make flat within the same folder?

**Comment by MattKintonCinch** [comment]
@jonaslagoni no worries, managed to get round our specific problem. This would still be a really good feature though :)

**Comment by jonaslagoni** [comment]
For anyone looking to get into this issue, we have described how to add new generators here: https://github.com/asyncapi/modelina/blob/master/docs/contributing.md#adding-a-new-generator

Always happy to help you get started, just ping me here or on slack 🙂

**Comment by akkshitgupta** [comment]
Is this issue still relevant? I would like to work on this.  

@jonaslagoni please lemme know if I should start working on this if not resolved already. Happy to assist in other tasks as well if someone is looking into it.

**Comment by jonaslagoni** [comment]
@akkshitgupta it is yes!

Here is how I would achieve it:

- Follow the guide for adding a new generator: https://github.com/asyncapi/modelina/blob/master/docs/contributing.md#adding-a-new-generator
- Adapt the template generator to output the desired JSON Schema output, this might be a bit tricky, but don't get discouraged, just reach out and I will help you if you get stuck!
